### PR TITLE
Take into account relativistic effects in Poisson solver

### PR DESCRIPTION
**Pull request by RemiLehe**
This PR modifies the Poisson solver from #465 in order to take into relativistic effects (i.e. when the particles that produce the space-charge field have a relativistic velocity.)

Summary of changes:
- In `InitSpaceChargeFields`, after depositing the charge density of the particles, their average velocity is also calculated. This is then passed to the routines that calculate the space-charge fields.
- When computing the potential `phi` (`computePhi`), we now use the modified Laplacian operator `MLNodeTensorLaplacian`, which was recently implemented by @WeiqunZhang in `amrex` in order to represent the relativistic Poisson equation (see the Doxygen string of `computePhi` for the mathematical expression of the relativistic Poisson equation)
- The computation of `E` from `phi` (`computeE`) has been modified in order to take relativistic effects into account (see the mathematical expression in Doxygen string).
- A new function `computeB` was added, in order to compute the magnetic field associated with a relativistic charge distribution.
- In practice, whenever the beam is very relativistic, the default `amrex` solver fails to converge to the default precision (`1.e-11`) within a reasonable number of iterations (200) ; this then causes the simulation to abort. In order to by-pass this problem, the user can set the required precision for the space-charge field to a lower value. (An alternative is to compile the code with the `hypre` solvers).
- I added a 3D automated test, in which the beam is relativistic (gamma=100)

### CODE REVIEW

File: Source/Initialization/InitSpaceChargeField.cpp [hunk 1]
```diff
@@ -120,25 +149,136 @@ WarpX::computeE(amrex::Vector<std::array<std::unique_ptr<amrex::MultiFab>, 3> >&
...
+            // Calculate the electric field
+            // Use discretized derivative that matches the staggering of the grid.
 #if (AMREX_SPACEDIM == 3)
             amrex::ParallelFor( tbx, tby, tbz,
                 [=] AMREX_GPU_DEVICE (int i, int j, int k) {
-                    Ex_arr(i,j,k) += -inv_dx*( phi_arr(i+1,j,k) - phi_arr(i,j,k) );
+                    Ex_arr(i,j,k) +=
+                        +(beta_x*beta_x-1)*inv_dx*( phi_arr(i+1,j,k)-phi_arr(i,j,k) )
+                        +beta_x*beta_y*0.5*inv_dy*( phi_arr(i,j+1,k)-phi_arr(i,j-1,k) )
+                        +beta_x*beta_z*0.5*inv_dz*( phi_arr(i,j,k+1)-phi_arr(i,j,k-1) );
```
Comment (WeiqunZhang) [inline Source/Initialization/InitSpaceChargeField.cpp]: For the dy and terms, I think, phi at i+1 need to be used too.

Comment (RemiLehe) [reply Source/Initialization/InitSpaceChargeField.cpp]: Ah, yes, that's right. I'll update the PR accordingly.

**Comment by RemiLehe** [comment]
@WeiqunZhang I took into account and updated this PR (sorry for the delay).
Would you have time to have another look?

File: Source/Initialization/InitSpaceChargeField.cpp [hunk 2]
```diff
@@ -251,34 +259,44 @@ WarpX::computeB (amrex::Vector<std::array<std::unique_ptr<amrex::MultiFab>, 3> >
...
+                        -beta_z*inv_dx*0.5*(phi_arr(i+1,j  ,k)-phi_arr(i,j  ,k)
+                                          + phi_arr(i+1,j+1,k)-phi_arr(i,j+1,k))
+                        +beta_x*inv_dz*0.5*(phi_arr(i  ,j+1,k)-phi_arr(i  ,j,k)
+                                          + phi_arr(i+1,j+1,k)-phi_arr(i+1,j,k)));
                 },
                 [=] AMREX_GPU_DEVICE (int i, int j, int k) {
                     Bz_arr(i,j,k) += inv_c * (
-                        +beta_y*inv_dx*( phi_arr(i+1,j,k)-phi_arr(i,j,k) ));
+                        +beta_y*inv_dx*0.5*(phi_arr(i+1,j  ,k)-phi_arr(i,j  ,k)
+                                          + phi_arr(i+1,j+1,k)-phi_arr(i,j+1,k)));
```
Comment (WeiqunZhang) [inline Source/Initialization/InitSpaceChargeField.cpp]: It seems that the original 2D code for B is correct.

Comment (RemiLehe) [reply Source/Initialization/InitSpaceChargeField.cpp]: Ah, yes, that's right - except for `By` which is cell-centered in both x and z. I'll update the PR.

**Pull request merged**